### PR TITLE
Asychronous RPC requests

### PR DESCRIPTION
**Pull request by jagerman**
This adds a generic interface for making asychronous RPC request
handlers, and applies these to the reward and exit signature endpoints.

The RPC interface adds an alternative the `invoke(...)` method to
optionally take a new, third argument of a `shared_ptr<response>`: when
such a version of the invoke method is present then response is not send
until the destruction of the shared_ptr, and so an asychronous request
uses these, keeps the shared_ptr alive until the response is available,
then lets it destruct which then triggers the response.

This then pushes the async approach through the reward and exit
signature requests, and gets everything working through asychronous
callbacks rather than blocking requests.

This also makes some small changes to signature handling:

- Allow exit/liquidation RPC requests to be made by either SN pubkey (as
  currently) or BLS pubkey (new).

- Allow liquidation of oxend-non-existent BLS pubkeys (i.e. liquidating
  a BLS pubkey that doesn't match any SNs oxen knows of); without this
  it isn't possible to remove a "bad" contract node that oxend didn't
  accept the registration of for whatever reason.

- Add code to remove unwanted node extra signatures from produced
  signatures.  Previously our concept of "non-signers" only included IDs
  to pass to the contract, but there is also a reverse failure were we
  collect a signature from a SN that isn't in the contract anymore (for
  example: a recently removed node with an incoming, but unconfirmed,
  exit event). This amends the signing code to detect any such signers
  and subtract the signatures of any such SNs from the aggregate before
  returning it.

- Removes the walk-the-snode-linked-list contract handling code as it is
  not used anymore.

### CODE REVIEW

File: src/bls/bls_aggregator.cpp
Comment (Doy-lee) [inline src/bls/bls_aggregator.cpp]:
```suggestion
            cryptonote::core& core, const std::unordered_map<bls_public_key, bls_signature>& signatures) {

        if (log::get_level(logcat) <= log::Level::debug)
            core.l2_tracker().get_all_service_node_ids(
                    std::nullopt,
                    [&core, signatures](
```

Take by ref and only produce a copy if in the lambda, e.g. in debug logging mode only.

File: src/bls/bls_aggregator.cpp
Comment (Doy-lee) [inline src/bls/bls_aggregator.cpp:549]: I'm wondering if we need this `MAX_CONNECTIONS`? Internally OMQ has a thread pool from my understanding, is there a need to rate limit how many jobs we dispatch? Maybe there's a queue limit in OMQ we have to respect, but if we omit this then we don't need the recursion which would simplify the algorithm.

If there was a queue limit though I feel like we should be querying that from OMQ rather than an ad-hoc check here (because multiple BLS requests will basically fill it up without us being able to rate limit ourselves)

Comment (Doy-lee) [reply src/bls/bls_aggregator.cpp:549]: On second review, I see that essentially the threads allocated for the RPC server are only for parsing and submitting the request into a queue. It returns immediately so that the handler can service more requests. It's then OMQ's responsibility to dequeue and submit the network request and this is done recursively using the thread allocated for the network request from their thread pool. The consequences of this is fairly benign because after the network request is done it recurses into `establish()` to dispatch more network requests and exits ASAP after that. 

The work load the thread has to do post-response (recursive `establish()` and friends) is fairly small so the window of time it actually blocks other network requests in OMQ waiting for the post-amble to finish is small.

This is the same pattern AFAICT is in L2Tracker, Ethyl but also the CURL revamp in the storage server. But I have issues in this pattern in that the control flow and the complexity for this is very high because there's a need to smuggle the response lifetime across multiple levels of callbacks and that asynchronous implementation gets leaked into the RPC implementation code.

The top level goal here being that _something_ has to release the RPC server from waiting for the response because it doesn't want to block the handling of other RPC request. That _something_ we're punting this responsibility to is the OMQ threads (but _only_ if the request requires recursion, e.g. the max connections > 900).

Is there a simpler solution here that we just spawn a thread to handle the RPC request, capture the response context into the thread and let it run detached? This abstracts the concept that the RPC implementation has to be aware that they may be operating in an asynchronous manner. RPC code can be written synchronously where it's then the invoker's responsibility to prep to execute the RPC code asynchronously. That drops the recursion and the helper data structure `rpc_responder` and auxiliary code to process that.

One of the characteristics we did gain from the current implementation is that the number of threads concurrently active by the process is bounded to the number of threads allocated to the RPC server. If we begin spawning a thread for each dispatched request we lose that but that is easily handled by just having some atomic counter to cap the number of threads spawned by that system.

Comment (jagerman) [reply src/bls/bls_aggregator.cpp:549]: The `MAX_CONNECTIONS` isn't about processing concurrency, but rather about the number of simultaneous TCP connections (and thus file descriptors) that we will open at a time to issue these requests.  900 is a bit under the typical 1024 file descriptor limit, and we close these connections immediately as we get responses returned so that even if we open new ones, we should stay at a reasonably safe number.

I do worry that 900 seems a bit high since any incoming HTTP RPC requests are *also* using a file descriptor; I think we could reasonably lower it to something like 300-500 without impacting total request time too much.

> The top level goal here being that something has to release the RPC server from waiting for the response because it doesn't want to block the handling of other RPC request. That something we're punting this responsibility to is the OMQ threads (but only if the request requires recursion, e.g. the max connections > 900).

That's not quite how this works: we are *always* (`*` almost) releasing the RPC thread from responsibility until we have accumulated every response from the remotes, regardless of the number of connections.  Even if there's just one request, we don't want to block waiting for that remote request to come back: and so we create the request state, initiate the connection and request on that connection, then return immediately, surrendering control.  Control resumes when a response comes back, at which point we continue processing, accumulating the single result, possibly establishing new connections if more are needed, or if we got the very last response, fulfilling the RPC response (and release our state).  That entire process could happen in a second, or might be a dozen seconds, depending on network latencies and timeouts.

`*` -- well, not quite always: if we serve a repeated request it from cache then there's no async control surrender at all.

> Is there a simpler solution here that we just spawn a thread to handle the RPC request, capture the response context into the thread and let it run detached?

The explicit intention of this PR is to get *away* from that.

> One of the characteristics we did gain from the current implementation is that the number of threads concurrently active by the process is bounded to the number of threads allocated to the RPC server. If we begin spawning a thread for each dispatched request we lose that but that is easily handled by just having some atomic counter to cap the number of threads spawned by that system.

Doing so would just be pushing the bottleneck to a different place, but you still end up with a limit based on uncontrollable external network latencies (and timeouts) rather than local resources, and you can still end up in a situation where some network event (in this case: waiting for timeouts from slow or offline nodes) can still back up your request queue to the point that it is prevent from accepting new requests, and then you have the issue of what to do if a signing request comes in and all your threads are currently blocked waiting on responses: do you reply with an RPC failure?  Do you push the blockage further back (effectively blocking other RPC requests)?  Neither of these are particularly desirable.

Using threads to make blocking requests feels like an anti-pattern to me; yes non-blocking code *is* more complicated in general (whatever the domain), but it's also considerably more efficient in that your internal code is idle until something triggers it with something to do, rather than having multiple blocked threads waiting on responses from some unpredictable external factor.

With this async approach, we have no practical limit (other than system resources) on how many requests we can handle: the only time any resources are occupied is when processing the rpc request, initiating outbound connections, or processing the response to requests on this connections, but otherwise resources are idle and available for other tasks.

This is also an issue we have to deal with in multiple places already, in storage-server and lokinet but *also* inside existing oxen-core http rpc request processing, just lower down: uWebSockets is asynchronous by design, where there just isn't a single request callback because multiple requests are in various states of delivering data at any given time.  All of these use a similar approach of stashing the ongoing state into an object inside a callback to get completed later.

Direct OMQ or QUIC requests make this much easier (you can just grab a `msg.send_later()` and store that until you're ready to reply), but the RPC interface in core -- where you don't *reply*, you just set up the response in an object already waiting for you -- made it a bit messier here.

Comment (Doy-lee) [reply src/bls/bls_aggregator.cpp:549]: > That's not quite how this works: we are always (* almost) releasing the RPC thread from responsibility until we have accumulated every response from the remotes, regardless of the number of connections. Even if there's just one request, we don't want to block waiting for that remote request to come back: and so we create the request state, initiate the connection and request on that connection, then return immediately, surrendering control. Control resumes when a response comes back, at which point we continue processing, accumulating the single result, possibly establishing new connections if more are needed, or if we got the very last response, fulfilling the RPC response (and release our state). That entire process could happen in a second, or might be a dozen seconds, depending on network latencies and timeouts.

Yep this is what I meant. As in control is relinquished from the RPC thread but it doesn't disappear, we moved control to the OMQ thread that is handling the request, when that eventually finishes the callback is triggered, on the OMQ thread that is serving the last request in the aggregation (correct me if I'm wrong, that thread is sleeping and blocked).

So to rephrase more aptly, it's a coroutine implemented using asynchronous lambdas that proceeds the BLS aggregation state machine. This has complexities as we've mentioned. The complexity is avoidable using a different pattern like a coordinating thread that selectively forwards the request to OMQ and waits for the response on the same thread.

> Doing so would just be pushing the bottleneck to a different place, but you still end up with a limit based on uncontrollable external network latencies (and timeouts) rather than local resources, and you can still end up in a situation where some network event (in this case: waiting for timeouts from slow or offline nodes) can still back up your request queue to the point that it is prevent from accepting new requests, and then you have the issue of what to do if a signing request comes in and all your threads are currently blocked waiting on responses: do you reply with an RPC failure? Do you push the blockage further back (effectively blocking other RPC requests)? Neither of these are particularly desirable.

So I'm identifying 2 problems I want to solve based off the discussion so far.

1) Is it possible to get rid of the async complexity from the code
2) How do we gracefully handle the limits of the system (new concern you've brought up)

A simple scenario that breaks the limits of our proposed systems. Submit 2x BLS aggregations on the same tick to the RPC endpoint. On mainnet with 2k nodes, a BLS aggregation will generate 1800 (max 900 per BLS request atm) requests to the OMQ queue. This will most likely break with the async/detached thread method because they act independently to each other and try to jam the queue without considering each other (and anecdotally I've seem OMQ freeze up with a 1k job limit, and yes we can bump this higher and move the can down the road a bit further).

> Using threads to make blocking requests feels like an anti-pattern to me; yes non-blocking code is more complicated in general (whatever the domain), but it's also considerably more efficient in that your internal code is idle until something triggers it with something to do, rather than having multiple blocked threads waiting on responses from some unpredictable external factor.

I'm not sure about this concern because we are blocking threads, we're blocking in OMQ, or whatever mechanism is in there to wait for the response to return. _Something_ is waiting for epoll/kqueue/io_uring, it's never free. It's that if we move the blocking out of OMQ into essentially userland code (Oxen) we have granular control over when they are forwarded into OMQ.

--

So for 1) initially my first draft was to suggest a detached thread because it's easy to implement that. That in itself removes the need for recursive lambdas because you have a thread that will sit around waiting for the response. 

But for 2) it's insufficient in actually controlling when the request is forwarded onto OMQ. The next step for that builds on that by actually having a coordinating thread that dispatches those jobs to OMQ. This is similar in to how io_uring works. 

The reason for that is you only want to dequeue requests when the underlying system, the coordinating thread _knows_ that OMQ has the capacity to serve the request. So requests come in, it stores them into a queue, RPC thread returns so it can await more requests, the requests gets dequeued and executed when it knows OMQ has capacity to handle it.

Then the amount of requests we can service is limited to the coordinating thread's queue size. It's tweakable, it can be scaled up, it can use a fixed amount of resources, it can rate limit how many requests are pushed into OMQ e.t.c.

Comment (jagerman) [reply src/bls/bls_aggregator.cpp:549]: > Yep this is what I meant. As in control is relinquished from the RPC thread but it doesn't disappear, we moved control to the OMQ thread that is handling the request, when that eventually finishes the callback is triggered, on the OMQ thread that is serving the last request in the aggregation (correct me if I'm wrong, that thread is sleeping and blocked).

There's a bit of a misunderstanding here about how the request handler works; with this PR, there is never any thread blocked on waiting on any part of an oxend request.  (There are, in a technical sense, low-level event loop threads underlying zeromq and uWebSocket, but they are basically only "blocked" when there is no IO nor timers that need them to wake up and do something, and so really more "idle" than "blocked").
1. An HTTP arrives asynchronously piecemeal via uWebSockets (which has its own event loop thread that only handles incoming and outgoing data); as each piece arrives, we accumulate it; when we have the entire request we then queue a job (via an oxenmq thread) to process it via oxend rpc request handling.  The uWebSocket thread, however, does not wait on this request handler (not even for it to start; it just queues and moves on).
- The oxenmq job is processed by some available OMQ worker thread, usually pretty much instantly unless there aren't any available workers in which case it gets queued for processing as soon as a worker frees up.
- For _non-asynchronous requests_, the worker thread invokes the RPC method, which is expected to fill the response field of the RPC request object passed into it, then serializes that response (typically to json), and queues it back into uWebSockets to be sent back to the HTTP requestor when the uWebSockets loop is able (which is pretty much instantly, but just pointing out that it's still a thread-to-thread transfer of the data).  OMQ responses for OMQ RPC requests are similar.  Queuing these responses does block in either case.  As soon as the response are handed off, the OMQ-managed thread that was processing the job up until this point goes back to look for jobs (or else idles until something becomes available).
    - zeromq or uWebSockets is then responsible for sending out the message to the recipient.  (Both use non-blocking I/O to do so).
- For _async requests_, added in this PR, the process is similar up to the the RPC method, but then it differs: that method is called with an instantiated `std::shared_ptr<responder>`.  That object has two purposes: it keeps the request state alive, and it allows fulfilling the request response at some later point (or, if it gets destroyed without responding, sends an error response).  The RPC method in question stores this shared pointer somewhere to keep the request data alive, initiates some non-instant action, and returns immediately.  The request is then just abandoned by the worker processing this part of the request because it has transferred that responsibility to the request handling method it called.
  - (as a special case, if we already have the result in cache, we *don't* store that response object but rather fulfill it immediately during the rpc handler method)
  - at some point "something happens" that triggers us having sufficient data to make the response (that's a generic description; in this PR that trigger is when we have accumulated every response and/or timeout to our dispatched SN requests), and this triggers us to use our stored `std::shared_ptr<responder>` to fulfill the response.
  - that request fulfillment happens just as the non-async request fulfillment would, it's just being as another worker job, likely on a different thread.

So there's no thread anywhere is occupied with this request except when actively doing something with it.  In between event triggers state is held via lambda captures, but there is no dedicated thread here anywhere in the process, there is only dedicated *data* that is managed by whichever thread ends up gathering enough data to send the final response.

> I'm not sure about this concern because we are blocking threads, we're blocking in OMQ, or whatever mechanism is in there to wait for the response to return.

There is no such blocking here.  (There is an idle thread that will be woken up when *any* incoming data arrives, or a timer arrives, but that really is idle rather than blocking because nothing we've done prevents it from doing other work.  It'll get woken up when packets contain our response arrive, but it'll also get woken up and process a myriad other things that might happen before then).

> Something is waiting for epoll/kqueue/io_uring, it's never free. It's that if we move the blocking out of OMQ into essentially userland code (Oxen) we have granular control over when they are forwarded into OMQ.

Moving it out of OMQ doesn't *remove* any blocking/sleeping code, it just adds *another* blocking thread that still has to wait on events to fire in response to incoming data from the existing zeromq or uwebsocket internal thread, which wouldn't go away (and then a question of what it does with multiple simultaneous requests).  In other words, with this PR we have a flow like this, where "->" is handing data off to another thread:

- [uwebsockets] -> [worker thread processing request]
- [OMQ IO thread SN response] -> [worker thread accumulating response]
- [OMQ IO thread SN response] -> [worker thread accumulating response]
- ...
- [OMQ IO thread SN response] -> [response accumulator that gets last response] -> [uwebsockets response queue]

and if we have a "waiter" thread around it then we just get:
- [uwebsockets] -> [worker thread processing request] -> [waiter]
- [OMQ IO thread SN response] -> [worker thread] -> [waiter response accumulator]
- [OMQ IO thread SN response] -> [worker thread] -> [waiter response accumulator]
- ...
- [OMQ IO thread SN response] -> [worker thread] -> [waiter response last response] -> [uwebsockets response queue]

and so effectively all you do is shoehorn all the response handlers to pass the data off to another synchronous function that is going to end up looking pretty much exactly like `bls_aggregator.cpp`'s internal `nodes_request_data` class, but with a mutex+condition_variable feeding results in as they arrive instead of a mutex+callback handling the results as they arrive.

So you end up at the same place, but have added more threads with an extra layer of passing data between threads and quite possibly a new need for a thread pool to be able to manage multiple requests.

> But for 2) it's insufficient in actually controlling when the request is forwarded onto OMQ. The next step for that builds on that by actually having a coordinating thread that dispatches those jobs to OMQ. This is similar in to how io_uring works.

Are you talking here about when multiple requests are active at once, and that we would be exceeding the MAX_CONNECTIONS (e.g. 900 on request 1, and possibly another 900 on request 2)?  That definitely is a concern, but doesn't need a dedicated thread to solve.  I will think about this some more, but I think it is quite easily solvable by having `nodes_request_data` have some shared state across requests.

Or we replace it all with libquic and forget about worrying about connection limits.  (Not a serious suggestion; as much as I'm tempted, there's enough transition underway already).

Comment (Doy-lee) [reply src/bls/bls_aggregator.cpp:549]: Yeah if OMQ and consequently ZMQ are reading data from non-blocking IO then it won't end up blocking a thread. Thanks for the explanation btw, it's very useful. Though we still lose control of the handle to that managed data unless we collect it and keep it around ergo:

>  Are you talking here about when multiple requests are active at once, and that we would be exceeding the MAX_CONNECTIONS (e.g. 900 on request 1, and possibly another 900 on request 2)? That definitely is a concern, but doesn't need a dedicated thread to solve. I will think about this some more, but I think it is quite easily solvable by having nodes_request_data have some shared state across requests.

Collecting them and tracking the concurrent ones is how I'd solve this quoted problem. At some point somewhere, something needs to know how much work has been sent to OMQ to be able to reason about the limits of the system. 

The typical usecase will be interacting with the network through our public staking portal which is going to point to at most couple of daemons if find we need some load balancing so it might be/might not be a plausible situation that it's receiving the aggregation requests within very close ticks to each other just from general public usage.

re: Thread vs callbacks+managed data, I'm on the thread side of this discussion because to me that affords us a dedicated stack frames that you can store things like the callbacks/managed data, cancel them and so forth.

So I think it's raised some interesting points. I will merge this and let this move on forward, I think maybe I've planted some ideas that might be useful in the future but I'm not going to block this further.

I will hold on merging this until the syncing of dev onto stable is fixed as I don't want to add more code to the debugging that's going on at the moment.

**Review by Doy-lee — APPROVED** [review]

**Pull request merged**
